### PR TITLE
Special cased Go's PutObject request handler

### PR DESCRIPTION
**Pull request by RachelTucker**
Generated code:
```
package models

import (
    "net/url"
    "net/http"
    "ds3/networking"
    "strings"
    "strconv"
)

const ( AMZ_META_HEADER = "x-amz-meta-" )

type PutObjectRequest struct {
    bucketName string
    objectName string
    checksum networking.Checksum
    content networking.ReaderWithSizeDecorator
    headers *http.Header
    job string
    offset int64
    queryParams *url.Values
}

func NewPutObjectRequest(bucketName string, objectName string, content networking.ReaderWithSizeDecorator) *PutObjectRequest {
    queryParams := &url.Values{}

    return &PutObjectRequest{
        bucketName: bucketName,
        objectName: objectName,
        content: content,
        checksum: networking.NewNoneChecksum(),
        headers: &http.Header{},
        queryParams: queryParams,
    }
}

func (putObjectRequest *PutObjectRequest) WithJob(job string) *PutObjectRequest {
    putObjectRequest.job = job
    putObjectRequest.queryParams.Set("job", job)
    return putObjectRequest
}
func (putObjectRequest *PutObjectRequest) WithOffset(offset int64) *PutObjectRequest {
    putObjectRequest.offset = offset
    putObjectRequest.queryParams.Set("offset", strconv.FormatInt(offset, 10))
    return putObjectRequest
}



func (PutObjectRequest) Verb() networking.HttpVerb {
    return networking.PUT
}

func (putObjectRequest *PutObjectRequest) Path() string {
    return "/" + putObjectRequest.bucketName + "/" + putObjectRequest.objectName
}

func (putObjectRequest *PutObjectRequest) QueryParams() *url.Values {
    return putObjectRequest.queryParams
}

func (putObjectRequest *PutObjectRequest) WithChecksum(contentHash string, checksumType networking.ChecksumType) *PutObjectRequest {
    putObjectRequest.checksum.ContentHash = contentHash
    putObjectRequest.checksum.Type = checksumType
    return putObjectRequest
}

func (putObjectRequest *PutObjectRequest) GetChecksum() networking.Checksum {
    return putObjectRequest.checksum
}

func (putObjectRequest *PutObjectRequest) WithMetaData(key string, value string) *PutObjectRequest {
    if strings.HasPrefix(strings.ToLower(key), AMZ_META_HEADER) {
        putObjectRequest.headers.Add(strings.ToLower(key), value)
    } else {
        putObjectRequest.headers.Add(strings.ToLower(AMZ_META_HEADER + key), value)
    }
    return putObjectRequest
}

func (putObjectRequest *PutObjectRequest) Header() *http.Header {
    return putObjectRequest.headers
}

func (putObjectRequest *PutObjectRequest) GetContentStream() networking.ReaderWithSizeDecorator {
    return putObjectRequest.content
}
```

### CODE REVIEW

**Comment by rpmoore** [comment]
Does this allow a user to specify metadata when putting the object?

**Comment by RachelTucker** [comment]
No it does not. I will update

File: ds3-autogen-go/src/main/resources/tmpls/go/request/with_headers.ftl [hunk 1]
```diff
@@ -0,0 +1,12 @@
+func (${name?uncap_first} *${name}) WithMetaData(key string, value string) *${name} {
+    if strings.HasPrefix(strings.ToLower(key), AMZ_META_HEADER) {
+        ${name?uncap_first}.headers.Add(key, value)
```
Comment (rpmoore) [inline ds3-autogen-go/src/main/resources/tmpls/go/request/with_headers.ftl]: Does the headers.Add method `ToLower` the keys?

Comment (RachelTucker) [inline ds3-autogen-go/src/main/resources/tmpls/go/request/with_headers.ftl]: No it does not lower case them. If I travel through several layers of interfaces I get the documentation
```
// CanonicalMIMEHeaderKey returns the canonical format of the
// MIME header key s. The canonicalization converts the first
// letter and any letter following a hyphen to upper case;
// the rest are converted to lowercase. For example, the
// canonical key for "accept-encoding" is "Accept-Encoding".
// MIME header keys are assumed to be ASCII only.
// If s contains a space or invalid header field bytes, it is
// returned without modifications.
```

Comment (rpmoore) [inline ds3-autogen-go/src/main/resources/tmpls/go/request/with_headers.ftl]: It would be safer to lowercase them in this method then.

Comment (RachelTucker) [inline ds3-autogen-go/src/main/resources/tmpls/go/request/with_headers.ftl]: So lower case them in our code, and let the networking library canonicalize them?

Comment (rpmoore) [inline ds3-autogen-go/src/main/resources/tmpls/go/request/with_headers.ftl]: I believe this is what we do in the other sdks.

Comment (RachelTucker) [inline ds3-autogen-go/src/main/resources/tmpls/go/request/with_headers.ftl]: will update

**Pull request merged**